### ISSUE TITLE
Release & document Chroma

### ISSUE THREAD

**Issue by jakubfiala** [meyda/meyda#218]
This should happen following the merging of #215 and #216 

we need to:

- [x] check if any API-breaking changes were made (probably not)
- [x] update the package version accordingly
- [x] update docs, mainly [this page](https://github.com/meyda/meyda/wiki/audio-features)
- [x] do `npm publish` and a GitHub release
- [ ] celebrate by measuring the chroma of ALL the things 🎉

**Comment by jakubfiala** [comment]
gonna self-assign this for now but I'd appreciate any help, mainly figuring out what needs to be in the docs @odub @meyda/owners

**Comment by jakubfiala** [comment]
### New version number

so it looks like the hop size addition doesn't break API because if not specified, it just quietly falls back to `bufferSize` – if HS===BS then the output should be identical to v4.0.5

The Chroma addition doesn't break API either, of course

I suggest we release this as **meyda@4.1.0**, lmk if you have any objections. 

### Docs

I'm going to update the wiki page listing available features, adding chroma as a spectral feature. We could think about having a separate section for tonal features, but since chroma is currently the only one, I think it could totally stay in spectral.

I'm going to cite https://labrosa.ee.columbia.edu/matlab/chroma-ansyn/chromagram_E.m as the source implementation (as `librosa` also references it).

As feature description, I was going to write:

```
Calculates the spectral energy of the signal for each of the chromatic pitch classes (C, C♯, D, D♯, E ,F, F♯, G, G♯, A, A♯, B). 
Often used for tonal audio analysis including pitch detection.
```

Does that sound good? 

For hop size, I'm going to just update the [Getting started page](https://github.com/meyda/meyda/wiki/Getting-Started) adding `hopSize` to the options object and explaining what it means below.

### Release

Provided that everyone's happy with the above, I can press the button tomorrow evening (Dec 3)

cc @meyda/owners @odub

**Comment by odub** [comment]
Hey Jakub,

I'm really excited by this.

I propose this slight tweak to the documentation, to clarify what's meant by tonal audio analysis:
```
Calculates the spectral energy of the signal for each chromatic pitch class (C, C♯, D, D♯, E, F, F♯, G, G♯, A, A♯, B). Often used to analyse the harmonic content of recorded music, such as in chord or key detection.
```

Happy with the plan!

**Comment by odub** [comment]
Will https://meyda.github.io get updated, by the way?

**Comment by jakubfiala** [comment]
> I propose this slight tweak to the documentation, to clarify what's meant by tonal audio analysis:

that sounds good, will use this!

re: website – you've got an updated version of the demo that you made the gif in the PR with, right? could we perhaps just use that?

**Comment by odub** [comment]
@jakubfiala Made #220 in case you want it. No worries either way though.

**Comment by jakubfiala** [comment]
awesome, thanks! it's really cool!

**Comment by nevosegal** [comment]
Good stuff! 🥇 Yeah, if this doesn't break anything, we should release it as `4.1.0` 👍

**Comment by jakubfiala** [comment]
damnit, I don't have permission to publish to the meyda org on npm!

@hughrawlinson are you the admin of the org? could you please add me? (or just run npm publish)

**Comment by jakubfiala** [comment]
I'll make a github release for now at least

**Comment by nevosegal** [comment]
I think I can do it. I just pulled the latest master, should I run it?

**Comment by jakubfiala** [comment]
awesome! yeah please do, the npm version + git tag is bumped up already :)

**Comment by nevosegal** [comment]
I tried publishing and it failed because my `gh-pages` wasn't up to date. I pulled the latest of that branch and ran the command again. Now it tells me that `npm ERR! You cannot publish over the previously published version 4.1.0. : meyda`... That's annoying. Though it looks like `npmjs.com` has version `4.1.0`.. so I'm confused

**Comment by jakubfiala** [comment]
oh wow, really? that's weird, maybe it worked for me but still threw an error saying i'm not allowed to publish … or is there perhaps a webhook of some sort that auto-publishes github releases?

**Comment by jakubfiala** [comment]
well, looks like it's published though :D

**Comment by nevosegal** [comment]
Yeah, I just hope this build isn't broken or something..

**Comment by jakubfiala** [comment]
hmm well Travis built it just fine, let me just quickly install it from npm and see if it works

**Comment by jakubfiala** [comment]
ok so it's actually broken, it didn't publish the `dist` directory 😧

**Comment by jakubfiala** [comment]
did you publish from `master`?

**Comment by nevosegal** [comment]
yes of course. Doesn't it build `dist` during `npm publish`? why the hell does it publish if it gives me a failed error...?!

**Comment by jakubfiala** [comment]
it does build, there's a prepublish hook – i really don't understand! i also didn't check if it published when i tried it or only after you did. Could you perhaps try doing `npm version 4.1.1` and publishing again?

**Comment by jakubfiala** [comment]
wait, I think i know what's going on, I think travis is publishing it automatically

**Comment by jakubfiala** [comment]
but it doesn't run `npm run build`, even though we specify it in `travis.yml`

**Comment by nevosegal** [comment]
hmm but npmjs.com says that I was the one who published it, around 15 min ago

**Comment by nevosegal** [comment]
I'm gonna try again when travis finishes building

**Comment by nevosegal** [comment]
ok, now I ran `npm install` again,`npm run build`, and pulled the latest of everything. Once #221 finishes I'll merge, pull and try to republish

**Comment by nevosegal** [comment]
Alright, that worked 👍 💯

**Comment by jakubfiala** [comment]
ok sweet

**Comment by jakubfiala** [comment]
amazing!

**Comment by jakubfiala** [comment]
ok, weirdly, if I npm install meyda and do `ls node_modules/meyda`, there's still no `/dist`, and I can't require the module

**Comment by jakubfiala** [comment]
could it be something to do with what's described in this file? https://github.com/meyda/meyda/blob/master/.npmignore

`/dist` is indeed in gitignore, so maybe npm is ignoring it now?

**Comment by odub** [comment]
This seems suspicious:
https://travis-ci.org/meyda/meyda/jobs/311589053#L1926

**Comment by jakubfiala** [comment]
ah, good spot @odub 
are we seriously not listing gulp in devdependencies?
how did this package ever have a successful build? :D

**Comment by jakubfiala** [comment]
oh no, we are listing it, but Travis isn't installing it for some reason

**Comment by 2xAA** [comment]
No, it's there 🤔 
https://github.com/meyda/meyda/blob/master/package.json#L59

**Comment by jakubfiala** [comment]
ok so I think what's happening is that Travis runs `npm run default` as the `script`, and as part of that also does `npm install`, but then it moves node_modules into the cache, and starts running the `deploy` job, which is separate and only runs `npm run build`, at which point gulp is no longer there.

I think what needs to happen is either we restore the dependency cache before `deploy`, or run npm install again.

I'm gonna sit down over lunch and fix travis, as well as move onto `babel-preset-env` to fix #222

**Comment by 2xAA** [comment]
That's confusing behaviour. If you don't get on to #222, I'll look into that soon.

**Comment by jakubfiala** [comment]
Yeah, I think the travis config could do with a cleanup – I really like the way CircleCI 2.0 does workflows, where you specify individual jobs and can move the cached files between them easily. 

What we've got is slightly strange. I also think we don't really need a deploy step, we should really be deploying manually to make sure things don't get released inadvertently.

**Comment by odub** [comment]
I wonder if the current problem could relate to the fragile-looking explicit paths in your npm scripts like this: https://github.com/meyda/meyda/blob/master/package.json#L14

It seems unnecessary to be explicit about the exact path, when you can just say `gulp build` or whatever and npm can handle it. Possibly this could mean that when the cache step happens on travis the cli tool gets lost. Maybe travis does some magic to add the cache directory to the PATH that the specific path to gulp breaks.

**Comment by jakubfiala** [comment]
@odub yeah that could well be it – I think I'll just get rid of the deploy step altogether to make things simpler

**Comment by jakubfiala** [comment]
@nevosegal I think I've fixed the CI stuff and bumped the version to 4.1.2 on master, any chance you could try `npm publish` again?

**Comment by nevosegal** [comment]
Published successfully

**Comment by jakubfiala** [comment]
well, the problem persists. The dist directory still doesn't appear. I really don't understand what's going on. We're always publishing locally, so how come the `/dist` directory isn't being picked up?

**Comment by odub** [comment]
Discovered today that you can run `npm pack` if you want to do a dry run of a publish

**Comment by odub** [comment]
Also this field might help: https://docs.npmjs.com/files/package.json#files

Totally weird that it worked previously but not any more :/

**Comment by jakubfiala** [comment]
hmm so I just tried running `npm pack` and `/dist` is there! wtf?! 
doesn't help that I still don't have the perms to publish to npm :/

@meyda/owners could anyone who has the perms try to fix this or give me the perms? Make sure you have the latest npm, I was also gonna try to remove `/dist` from gitignore and see if that fixes the issue

**Comment by odub** [comment]
+1 and @nevosegal I'd be curious to hear what npm version you're using.

**Comment by odub** [comment]
@jakubfiala: Apparently @hughrawlinson just fixed you up with the relevant permissions ✨ ✨

**Comment by jakubfiala** [comment]
yup, i'm on it

**Comment by jakubfiala** [comment]
@odub @nevosegal @2xAA @hughrawlinson this should now be fixed and published as `4.1.3` 🎉

The issue was that the 'empty npmignore' trick didn't work, i.e. npm was falling back to gitignore when publishing, but not when running `npm pack`. We obviously want to gitignore `/dist` but we don't want to npmignore it, so I had to add `!dist` to npmignore, which fixed the issue.

I will also close this issue as chroma is now out there.
